# 2018-12-05

- 몇 달 전에 배우면서 어려웠던 부분을 다시 정리하여 업로드 함

## Javascript
Async 함수는 항상 프라미스를 반환한다.

```javascript
async function f () {
  return 'Hello!'
}

console.log(f()) // Promise{}
```

Await 키워드는 대상이 Promise인 경우 해당 작업이 완료될 때 까지 대기하고, 대상이 Promise가 아닌 경우 기본 함수처럼 진행한다.

```javascript
function sum (a, b) {
  return a + b
}

function asyncSum (a, b) {
  return new Promise((resolve, reject) => {
    if (a === 0 && b === 0) {
      reject(new Error('I dont like 0'))
    } else {
      resolve(a + b)
    }
  })
}

async function test () {
  console.log(await sum(1, 3)) // 4, 그냥 진행
  console.log(await asyncSum(1, 3)) // 4, resolve 후 진행
}

test()
```

Async 함수에서 반환한 값은 Resolve된 Promise가 반환 된다. (Error를 Throw하면 Reject된 Promise 반환)
```javascript
async function hello () {
  return 'hello'
}

hello().then(r => console.log(r + '!!!')) // hello!!!
```


## Node.js

Node.js 프로세스가 알 수 없는 예외로 죽는것을 방지
```javascript
process.on('uncaughtException', e => {
  // Handling..
})
```